daffodil-runtime1/src/main/scala/org/apache/daffodil/runtime1/layers/api/Layer.scala

```diff
@@ -0,0 +1,275 @@
+ * as the single unique value of the layer, such as when the layer computes a
+ * checksum. This is, however, just a convention used by various checksum layers
+ * and classes/traits that support writing checksum layers.
+ *
+ * @param prefix preferred namespace prefix for the namespace of the variables
+ * @param namespace namespace of the variables
+ * @param variables list of pairs, each is the name (an NCName, that is without any
+ *                  namespace prefix) and type of a variable
+ */
+final case class LayerVariables(
+  prefix: String,
+  namespace: String,
+  variables: java.util.List[(String, DFDLPrimType)],
+)
+
+/**
+ * This is the primary API class for writing layers.
+ *
+ * All layers are derived from this class, and must have no-args default constructors.
+ *
+ * Derived classes will be dynamically loaded by Java's SPI system.
+ * The names of concrete classes derived from Layer are listed in a resources/M.services file
+ * so that they can be found and dynamically loaded.
+ *
+ * The SPI creates an instance the class of which is used as a factory to create the
+ * instances actually used by the Daffodil runtime. Compilation of the static information about
+ * the layer occurs only once and is then shared by all runtime instances.
+ *
+ * Instances of derived layer classes can be stateful. They are private to threads, and each time a layer
+ * is encountered during parse/unparse, an instance is created for that situation.
+ *
+ * All the static information about the layer is provided in the arguments.
+ *
+ * The rest of the Layer class implements the
+ * layer decode/encode logic, which is done as part of deriving one's Layer class from the
+ * Layer base class.
+ *
+ * About variables: Layer logic may read and write variables. Variables being read are parameters to
+ * the layer algorithm. Variables being written are outputs (such as checksums) from the layer algorithm.
+ * Variables being written must be undefined, since variables in DFDL are single-assignment.
+ * Variables being read must be defined before being read by the layer, and this is true for both
+ * parsing and unparsing. When unparsing, variables being read cannot be forward-referencing to parts
+ * of the DFDL infoset that have not yet been unparsed.
+ *
+ * @param layerName the name that will appear in the DFDL schema to identify the layer
+ * @param supportedLayerLengthKinds list of the layer length kinds the layer supports
+ * @param supportedLayerLengthUnits list of the layer length units the layer supports
+ * @param isRequiredLayerEncoding true if the layer is textual and so needs the layerEncoding property
+ * @param optLayerVariables a LayerVariables structure describing the variables the layer accesses
+ */
+abstract class Layer(
+  val layerName: String,
+  val supportedLayerLengthKinds: java.util.List[JLayerLengthKind],
+  val supportedLayerLengthUnits: java.util.List[JLayerLengthUnits],
+  val isRequiredLayerEncoding: Boolean,
```

Comment:
Is layer encoding needed anymore? It looks like CheckDigit and AIS are the only ones that set this to true. I don't see AIS using the layerCharset. CheckDigit layer uses it, but maybe we can get away with it just always being ascii?

Or alternatively, maybe this this is just another variable? For example, checkDigit already has a `checkDigitParams` variable, maybe an additional variable could be used to specify encoding?

Reply:
I think this can just be a DFDL variable, if a layer needs it then it can define a variable to use.